tests/test_ipython_magics.py

```diff
@@ -0,0 +1,193 @@
+        ),
+    ]
+
+
+def _ignore_cells_toml_input() -> List[Tuple[str, Callable[..., bool], str]]:
+    """Input for ignore cells test case using pyproject.toml configuration."""
+    toml_config_file = "pyproject.toml"
+    return [
+        (
+            '[tool.nbqa.ignore_cells]\nflake8=["%%custommagic"]',
+            _validate_ignore_cells_with_warning,
+            toml_config_file,
+        ),
+        (
+            '[tool.nbqa.ignore_cells]\nflake8=["%%custommagic", "%%anothercustommagic"]',
+            _validate_ignore_cells_without_warning,
+            toml_config_file,
+        ),
+    ]
+
+
+@pytest.mark.parametrize(
+    "config, validate, config_file",
+    [
+        *_ignore_cells_cli_input(),
+        *_ignore_cells_ini_input(),
+        *_ignore_cells_toml_input(),
+    ],
+)
+def test_ignore_cells(
+    config: str,
+    validate: Callable[..., bool],
+    config_file: Optional[str],
+    tmpdir: "LocalPath",
+    capsys: "CaptureFixture",
+) -> None:
+    """Validate we can ignore custom cell magics configured via cli, .ini and .toml."""
+    test_nb_path = _copy_notebook(
+        Path("tests/data/notebook_with_other_magics.ipynb"), Path(tmpdir)
+    )
+    nbqa_args = ["flake8", str(test_nb_path)]
+
+    if config_file:
+        _create_ignore_cell_config(Path(tmpdir) / config_file, config)
+    else:
+        nbqa_args.append(config)
+
+    with pytest.raises(SystemExit):
+        main(nbqa_args)
+
+    out, _ = capsys.readouterr()
+    assert validate(out, test_nb_path)
+
+
+def _indented_magics_toml_input() -> List[Tuple[str, Callable[..., bool], str]]:
+    """Input for handling indented magics test case using toml configuration."""
+    toml_config_file = "pyproject.toml"
+
+    return [
+        (
+            "[tool.nbqa.mutate]\nblack=1",
+            _validate_indented_magics,
+            toml_config_file,
+        )
+    ]
+
+
+def _validate_indented_magics(actual: str, test_nb_path: Path) -> bool:
+    """Validate the results of the notebook with indented magics after running black."""
+    expected_cell_source = [
+        ["import time\n", "\n", 'print(f"current_time: {time.time()}")'],
+        [
+            "from random import randint\n",
+            "\n",
+            "if randint(5, 10) > 8:\n",
+            '    %time print("Hello world")',
+        ],
+        ["str.split??"],
+        ["# indented line magic\n", "?str.splitlines"],
+        ["%time randint(5, 10)"],
+    ]
+    result: bool = True
+
+    with open(str(test_nb_path)) as handle:
+        actual_cells = json.load(handle)["cells"]
+        actual_cell_source = (cell["source"] for cell in actual_cells)
+
+    for actual_src, expected_src in zip(actual_cell_source, expected_cell_source):
+        result = result and (actual_src == expected_src)
+
+    return result and (actual == "")
+
+
+@pytest.mark.parametrize(
+    "config, validate, config_file",
+    [*_indented_magics_toml_input()],
+)
+def test_indented_magics(
```

Comment:
nice cleanup here!

Reply:
There is still some code duplication(code between the test case functions are similar). But I didnt want to go to extreme refactoring. If we are finding a pattern out of this, then we can further refactor.